File: app/src/main/java/com/grka/xray/config/JsonSubscriptionParser.kt
package com.grka.xray.config

import com.grka.xray.data.Profile
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.int
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

/**
 * Parses an xray-json / v2ray-json subscription body (as served by panels like
 * Remnawave) into profiles plus an optional routing template.
 *
 * Supported shapes:
 *  - a full config object: { "outbounds": [...], "routing": {...} }
 *  - an array of full configs: [ { "outbounds": [...], "routing": {...} }, ... ]
 *  - an array of bare outbound objects: [ { "protocol": ..., "settings": ... }, ... ]
 */
object JsonSubscriptionParser {

    data class Result(val profiles: List<Profile>, val routingJson: String?)

    private val json = Json { ignoreUnknownKeys = true }

    private val proxyProtocols = setOf("vless", "vmess", "trojan", "shadowsocks")

    fun looksLikeJson(body: String): Boolean {
        val t = body.trim()
        return t.startsWith("{") || t.startsWith("[")
    }

    fun parse(body: String): Result? = try {
        parseInner(body)
    } catch (e: Exception) {
        null
    }

    private fun parseInner(body: String): Result? {
        val root = json.parseToJsonElement(body.trim())

        val profiles = mutableListOf<Profile>()
        var routing: String? = null

        // The whole config is kept per profile so the panel's routing template
        // (with its original outbound tags) can be applied verbatim.
        fun handleConfig(obj: JsonObject) {
            val outbounds = obj["outbounds"]?.let { it as? JsonArray } ?: return
            if (routing == null) {
                obj["routing"]?.let { routing = it.toString() }
            }
            val fullConfig = obj.toString()
            // xray-json (Happ/Remnawave) carries the friendly server name here.
            val remarks = obj["remarks"]?.jsonPrimitive?.contentOrNull?.takeIf { it.isNotBlank() }
            var idx = 0
            for (ob in outbounds) {
                val outbound = ob as? JsonObject ?: continue
                val proto = outbound["protocol"]?.jsonPrimitive?.contentOrNull
                if (proto !in proxyProtocols) continue
                outboundToProfile(outbound, idx)?.let {
                    it.fullConfig = fullConfig
                    it.proxyTag = outbound["tag"]?.jsonPrimitive?.contentOrNull
                    if (remarks != null) {
                        it.name = if (idx == 0) remarks else "$remarks · ${it.proxyTag ?: (idx + 1)}"
                    }
                    profiles.add(it); idx++
                }
            }
        }

        when (root) {
            is JsonObject -> {
                when {
                    root.containsKey("outbounds") -> handleConfig(root)
                    root["protocol"] != null -> outboundToProfile(root, 0)?.let { profiles.add(it) }
                    else -> return null
                }
            }

            is JsonArray -> {
                var idx = 0
                for (item in root) {
                    val obj = item as? JsonObject ?: continue
                    when {
                        obj.containsKey("outbounds") -> handleConfig(obj)
                        obj["protocol"] != null -> outboundToProfile(obj, idx)?.let { profiles.add(it); idx++ }
                    }
                }
            }

            else -> return null
        }

        if (profiles.isEmpty()) return null
        return Result(profiles, routing)
    }

    private fun outboundToProfile(outbound: JsonObject, index: Int): Profile? {
        val protocol = outbound["protocol"]?.jsonPrimitive?.contentOrNull ?: return null
        if (protocol !in proxyProtocols) return null
        val settings = outbound["settings"] as? JsonObject ?: return null

        val p = Profile(protocol = protocol)

        when (protocol) {
            "vless", "vmess" -> {
                val vnext = (settings["vnext"] as? JsonArray)?.firstOrNull() as? JsonObject ?: return null
                p.server = vnext.str("address") ?: return null
                p.port = vnext.int("port") ?: return null
                val user = (vnext["users"] as? JsonArray)?.firstOrNull() as? JsonObject ?: return null
                p.uuid = user.str("id") ?: return null
                if (protocol == "vless") {
                    p.method = user.str("encryption") ?: "none"
                    p.flow = user.str("flow")
                } else {
                    p.method = user.str("security") ?: "auto"
                }
            }

            "trojan", "shadowsocks" -> {
                val server = (settings["servers"] as? JsonArray)?.firstOrNull() as? JsonObject ?: return null
                p.server = server.str("address") ?: return null
                p.port = server.int("port") ?: return null
                p.uuid = server.str("password") ?: return null
                if (protocol == "shadowsocks") p.method = server.str("method")
            }
        }

        applyStreamSettings(p, outbound["streamSettings"] as? JsonObject)

        // Preserve the exact outbound so non-standard transport fields (XHTTP
        // obfuscation, extra, noSSEHeader, xmux, …) survive verbatim.
        p.rawOutbound = outbound.toString()

        val tag = outbound["tag"]?.jsonPrimitive?.contentOrNull
        p.name = tag?.takeIf { it.isNotBlank() && it != "proxy" && it != "out" }
            ?: "${p.server}:${p.port}"
        return p
    }

    private fun applyStreamSettings(p: Profile, ss: JsonObject?) {
        if (ss == null) return
        p.network = ss.str("network") ?: "tcp"
        if (p.network == "http") p.network = "h2"

        when (p.network) {
            "ws" -> (ss["wsSettings"] as? JsonObject)?.let { ws ->
                p.path = ws.str("path")
                p.host = ws.str("host") ?: (ws["headers"] as? JsonObject)?.str("Host")
            }

            "xhttp" -> (ss["xhttpSettings"] as? JsonObject)?.let { xh ->
                p.path = xh.str("path")
                p.host = xh.str("host")
                p.xhttpMode = xh.str("mode")
                xh["extra"]?.let { if (it is JsonObject) p.xhttpExtra = it.toString() }
            }

            "grpc" -> (ss["grpcSettings"] as? JsonObject)?.let { g ->
                p.serviceName = g.str("serviceName")
                p.authority = g.str("authority")
                p.grpcMode = if (g["multiMode"]?.jsonPrimitive?.contentOrNull == "true") "multi" else null
            }

            "httpupgrade" -> (ss["httpupgradeSettings"] as? JsonObject)?.let { h ->
                p.path = h.str("path")
                p.host = h.str("host")
            }

            "h2" -> (ss["httpSettings"] as? JsonObject)?.let { h ->
                p.path = h.str("path")
                p.host = (h["host"] as? JsonArray)?.mapNotNull { it.jsonPrimitive.contentOrNull }?.joinToString(",")
            }

            "tcp" -> (ss["tcpSettings"] as? JsonObject)?.let { tcp ->
                val header = tcp["header"] as? JsonObject
                p.headerType = header?.str("type")
            }
        }

        p.security = ss.str("security")?.takeIf { it == "tls" || it == "reality" }
        val secObj = (ss["tlsSettings"] ?: ss["realitySettings"]) as? JsonObject
        secObj?.let { sec ->
            p.sni = sec.str("serverName")
            p.fingerprint = sec.str("fingerprint")
            p.allowInsecure = sec["allowInsecure"]?.jsonPrimitive?.contentOrNull == "true"
            p.alpn = (sec["alpn"] as? JsonArray)?.mapNotNull { it.jsonPrimitive.contentOrNull }?.joinToString(",")
            p.publicKey = sec.str("publicKey")
            p.shortId = sec.str("shortId")
            p.spiderX = sec.str("spiderX")
        }
    }

    private fun JsonObject.str(key: String): String? =
        this[key]?.let { runCatching { it.jsonPrimitive.contentOrNull }.getOrNull() }?.takeIf { it.isNotBlank() }

    private fun JsonObject.int(key: String): Int? =
        this[key]?.let { runCatching { it.jsonPrimitive.intOrNull }.getOrNull() ?: it.jsonPrimitive.contentOrNull?.toIntOrNull() }
}

File: apple/GrKaXKit/Sources/GrKaXKit/JsonSubscriptionParser.swift
import Foundation

/// Parses an xray-json / v2ray-json subscription body (as served by panels like
/// Remnawave) into profiles plus an optional routing template.
///
/// Supported shapes:
///  - a full config object: `{ "outbounds": [...], "routing": {...} }`
///  - an array of full configs
///  - an array of bare outbound objects
public enum JsonSubscriptionParser {

    public struct Result: Equatable, Sendable {
        public let profiles: [Profile]
        public let routingJson: String?
    }

    private static let proxyProtocols: Set<String> = ["vless", "vmess", "trojan", "shadowsocks"]

    public static func looksLikeJson(_ body: String) -> Bool {
        let trimmed = body.trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.hasPrefix("{") || trimmed.hasPrefix("[")
    }

    public static func parse(_ body: String) -> Result? {
        guard let root = JSON.parse(body.trimmingCharacters(in: .whitespacesAndNewlines)) else {
            return nil
        }

        var profiles: [Profile] = []
        var routing: String?

        /// The whole config is kept per profile so the panel's routing template
        /// (with its original outbound tags) can be applied verbatim.
        func handleConfig(_ object: JSONObject) {
            guard let outbounds = object["outbounds"]?.arrayValue else { return }
            if routing == nil, let node = object["routing"] {
                routing = node.compact
            }
            let fullConfig = JSON.object(object).compact
            // xray-json (Happ/Remnawave) carries the friendly server name here.
            let remarks = object["remarks"]?.nonBlankString
            var index = 0
            for entry in outbounds {
                guard let outbound = entry.objectValue else { continue }
                guard let proto = outbound["protocol"]?.stringValue,
                      proxyProtocols.contains(proto) else { continue }
                guard var profile = outboundToProfile(outbound) else { continue }

                profile.fullConfig = fullConfig
                profile.proxyTag = outbound["tag"]?.stringValue
                if let remarks {
                    profile.name = index == 0
                        ? remarks
                        : "\(remarks) · \(profile.proxyTag ?? String(index + 1))"
                }
                profiles.append(profile)
                index += 1
            }
        }

        switch root {
        case .object(let object):
            if object.has("outbounds") {
                handleConfig(object)
            } else if object.has("protocol") {
                if let profile = outboundToProfile(object) { profiles.append(profile) }
            } else {
                return nil
            }

        case .array(let items):
            for item in items {
                guard let object = item.objectValue else { continue }
                if object.has("outbounds") {
                    handleConfig(object)
                } else if object.has("protocol") {
                    if let profile = outboundToProfile(object) { profiles.append(profile) }
                }
            }

        default:
            return nil
        }

        if profiles.isEmpty { return nil }
        return Result(profiles: profiles, routingJson: routing)
    }

    private static func outboundToProfile(_ outbound: JSONObject) -> Profile? {
        guard let proto = outbound["protocol"]?.stringValue, proxyProtocols.contains(proto) else {
            return nil
        }
        guard let settings = outbound["settings"]?.objectValue else { return nil }

        var profile = Profile(proto: proto)

        switch proto {
        case "vless", "vmess":
            guard let vnext = settings["vnext"]?.arrayValue?.first?.objectValue else { return nil }
            guard let address = vnext["address"]?.nonBlankString else { return nil }
            guard let port = vnext["port"]?.intValue else { return nil }
            guard let user = vnext["users"]?.arrayValue?.first?.objectValue else { return nil }
            guard let id = user["id"]?.nonBlankString else { return nil }
            profile.server = address
            profile.port = port
            profile.uuid = id
            if proto == "vless" {
                profile.method = user["encryption"]?.nonBlankString ?? "none"
                profile.flow = user["flow"]?.nonBlankString
            } else {
                profile.method = user["security"]?.nonBlankString ?? "auto"
            }

        case "trojan", "shadowsocks":
            guard let server = settings["servers"]?.arrayValue?.first?.objectValue else { return nil }
            guard let address = server["address"]?.nonBlankString else { return nil }
            guard let port = server["port"]?.intValue else { return nil }
            guard let password = server["password"]?.nonBlankString else { return nil }
            profile.server = address
            profile.port = port
            profile.uuid = password
            if proto == "shadowsocks" {
                profile.method = server["method"]?.nonBlankString
            }

        default:
            return nil
        }

        applyStreamSettings(&profile, outbound["streamSettings"]?.objectValue)

        // Preserve the exact outbound so non-standard transport fields (XHTTP
        // obfuscation, extra, noSSEHeader, xmux, …) survive verbatim. This is
        // the whole reason the JSON type keeps key order.
        profile.rawOutbound = JSON.object(outbound).compact

        let tag = outbound["tag"]?.nonBlankString
        profile.name = tag.flatMap { $0 == "proxy" || $0 == "out" ? nil : $0 }
            ?? "\(profile.server):\(profile.port)"
        return profile
    }

    private static func applyStreamSettings(_ p: inout Profile, _ ss: JSONObject?) {
        guard let ss else { return }
        p.network = ss["network"]?.nonBlankString ?? "tcp"
        if p.network == "http" { p.network = "h2" }

        switch p.network {
        case "ws":
            if let ws = ss["wsSettings"]?.objectValue {
                p.path = ws["path"]?.nonBlankString
                p.host = ws["host"]?.nonBlankString ?? ws["headers"]?["Host"]?.nonBlankString
            }

        case "xhttp":
            if let xh = ss["xhttpSettings"]?.objectValue {
                p.path = xh["path"]?.nonBlankString
                p.host = xh["host"]?.nonBlankString
                p.xhttpMode = xh["mode"]?.nonBlankString
                if let extra = xh["extra"], extra.objectValue != nil {
                    p.xhttpExtra = extra.compact
                }
            }

        case "grpc":
            if let g = ss["grpcSettings"]?.objectValue {
                p.serviceName = g["serviceName"]?.nonBlankString
                p.authority = g["authority"]?.nonBlankString
                // Compared as text, so both `true` and `"true"` count — which
                // is what `jsonPrimitive.content` does upstream, and panels
                // write it both ways.
                p.grpcMode = g["multiMode"]?.stringValue == "true" ? "multi" : nil
            }

        case "httpupgrade":
            if let h = ss["httpupgradeSettings"]?.objectValue {
                p.path = h["path"]?.nonBlankString
                p.host = h["host"]?.nonBlankString
            }

        case "h2":
            if let h = ss["httpSettings"]?.objectValue {
                p.path = h["path"]?.nonBlankString
                p.host = h["host"]?.arrayValue?.compactMap { $0.stringValue }.joined(separator: ",")
            }

        case "tcp":
            if let tcp = ss["tcpSettings"]?.objectValue {
                p.headerType = tcp["header"]?["type"]?.nonBlankString
            }

        default:
            break
        }

        p.security = ss["security"]?.nonBlankString.flatMap { $0 == "tls" || $0 == "reality" ? $0 : nil }
        if let sec = (ss["tlsSettings"] ?? ss["realitySettings"])?.objectValue {
            p.sni = sec["serverName"]?.nonBlankString
            p.fingerprint = sec["fingerprint"]?.nonBlankString
            p.allowInsecure = sec["allowInsecure"]?.stringValue == "true"
            p.alpn = sec["alpn"]?.arrayValue?.compactMap { $0.stringValue }.joined(separator: ",")
            p.publicKey = sec["publicKey"]?.nonBlankString
            p.shortId = sec["shortId"]?.nonBlankString
            p.spiderX = sec["spiderX"]?.nonBlankString
        }
    }
}
